### PR TITLE
Added semicolon issues and examples

### PR DESCRIPTION
**Pull request by phistuck**

### CODE REVIEW

**Comment by i0natan** [comment]
@phistuck Welcome! Great addition.

I would suggest including only the second example as it much more simple and makes the point. The second is a bit sophisiticated, took me 5 min to understand using a code editor...

You might add console.log to ensure the reader understood that the function will actually return undefined

```
 function foo() {
  return // Oops, ; is automatically inserted here. foo() returns undefined
    {baz: true};
}

console.log(foo()) //undefined
```

**Comment by BrunoScheufler** [comment]
@phistuck awesome, thanks for the contribution! I agree to @i0natan comment to ensure the ease of understanding what's happening though, maybe you could adjust that a bit. Otherwise, great so far!

**Comment by phistuck** [comment]
Feel free to change whatever you want, you can even not give me credit for it, I really do not care.
The whole point of this pull request is the first case (as evident by the added text), because the second case is more-or-less covered in [3.3](https://github.com/i0natan/nodebestpractices#-33-start-a-codeblocks-curly-braces-on-the-same-line) (just without a return-new-line example for why this style is the "best practice", with which I disagree), if you go to the stackoverflow.com link that is mentioned there.

If you have a better scenario for the first case, I am happy to revise it.

**Comment by i0natan** [comment]
@phistuck I want to stick to your intent, can you kindly clarify to me the importance of the 1st example (in plain word, I probably miss the point there due to my ignorance)? 

p.s. The credit is yours, the 2nd example is also great

Toda!

**Comment by phistuck** [comment]
The importance is that leaving out semicolons can render a statement incomplete as far as the engine parses it and depending on what you write next, can have very unexpected behavior (it continues the statement with unintended code). Another example -
```
var foo = {} // Automatic semicolon insertion
var name = 'me' // Automatic semicolon insertion
var bar = foo // No automatic semicolon insertion because the next code starts with [

[name] = [1, 2, 3];  // Bang.
// The engine understands you wrote var bar = foo[name] = [1, 2, 3].
// foo is now {me: [1, 2, 3]}.
// bar is [1, 2, 3].
// :O
```

**Comment by i0natan** [comment]
@phistuck I was aware of the ASI oddities but didn't imagine it could be that fuzzy... Just ran the foo[name] example and laughed my s out :]

The examples are great to illustrate the fuzziness of ASI, just need to simplify it a bit for the average reader. @BrunoScheufler Can you collaborate here just to make the examples a bit shorter and easier (i.e. make the point with a quick punch)? since the code patterns section is on the homepage, the real estate is a bit more precious there

**Comment by phistuck** [comment]
I agree that my examples are convoluted, unfortunately. In the old world (well, middle-world), immediately-invoked-function-expressions were more likely to be scattered around the code (either by hand, or as a result of concatenating files), so having -
```
element.offsetHeight // whoops, this is going to be called!

(function () {
 // stuff
}());
```
Was more imaginable.

File: README.md [hunk 1]
```diff
@@ -259,7 +259,20 @@ function someFunction()
...
 
-**Otherwise:** As seen in the previous section, JavaScript's interpreter automatically adds a semicolon at the end of a statement if there isn't one which might lead to some undesired results
+**Otherwise:** As seen in the previous section, JavaScript's interpreter automatically adds a semicolon at the end of a statement if there isn't one, or considers a statement as not ended where it should, which might lead to some undesired results
+
+### Code example
+```javascript
+let count = 1
+// ...
+count = 2
+(function () { /* ... */ }()); // Oops, it tries to call count(), but count is not a function
```
[the remark this answers is not in the document]
Comment (phistuck) [reply README.md]: Hehe, I am so used to adding semicolons that I had to put them in a no-semicolon example. :) Removed.

The comment change was a last minute change before submitting, I should have known better. Fixed.

**Comment by sagirk** [comment]
Welcome @phistuck! Thanks for adressing my nitpicks.

I have a few suggestions to make the examples even more simpler and clearer:
- In the first example, remove the reassignment to `count` for simplicity
- In the second example, return a string instead of an object for simplicity
- For both, add good and bad versions to make things clearer for the average reader

```javascript
// Do
const count = 2;
(function doSomething() {
  // do something amazing
}());

// Avoid — throws exception
const count = 2 // it tries to run 2(), but 2 is not a function
(function doSomething() {
  // do something amazing
}())

// Do
function foo() {
  return 'javascript is awesome';
}

// Avoid - returns `undefined` instead of the value on the next line
function foo() {
  return // a semi-colon is automatically inserted here - foo() returns undefined
    'javascript is awesome'
}
```

**Comment by phistuck** [comment]
@sagirk - thinking about this, if it would call a function that changes some state instead of just calling a number, it would be better, because 2() will be very discoverable.
So something like -
```
function increase() { ++i }
function decrease() { --i }
const action = increase // oops, no semicolon, increase() was just called! i was increased
(function () {
// do something awesome
}());
```

What do you think?

**Comment by sagirk** [comment]
@phistuck I think this new example is relatively closer to something that can happen in a real-world scenario.

On the other hand, it might (or might not) make it difficult to follow for beginners.

There's a trade-off to be chosen here.

@i0natan @js-kyle @BrunoScheufler You guys have been here longer and therefore more familiar with the intended audience than I am. Which one would you pick?

**Comment by i0natan** [comment]
@sagirk My 2 pennies to spare: seek for simplicity and real-world examples. When the two interfere - opt for the simple (and short) one

@phistuck noted we share the same residence country, do you attend local meetups, fb/whatsapp groups?

**Comment by phistuck** [comment]
@i0natan - not into meetups or Facebook (or Telegram. Or WhatsApp, but there is no escape from that one). I am a Hangouts person, as worse as it sounds (anti-mobile me). I am about to join a UI Telegram group soon, though.

**Comment by i0natan** [comment]
@phistuck The local Whatsapp group is really vibrant and constitute many smart folks, LMK if you wanna join

**Comment by sagirk** [comment]
Hey @phistuck, just one more thing. Sorry to have missed bringing this up earlier.

Considering how the second example is covered in [3.3](https://github.com/i0natan/nodebestpractices#-33-start-a-codeblocks-curly-braces-on-the-same-line) (perhaps not nearly as well, but it gets the point across), and that real-estate is a bit more [precious](https://github.com/i0natan/nodebestpractices/pull/235#issuecomment-417950046) on the homepage, can we drop the second example and just keep the first one?

It seems to be the perfect sweet spot between too much and too little.

I'm happy to land it after this change. Let me know if I can help in any way.

**Comment by phistuck** [comment]
@BrunoScheufler, @i0natan and @sagirk - I do not mind removing and changing whatever all of you want, but I would like to have a full consensus first. :)

**Comment by i0natan** [comment]
@phistuck LOL

@sagirk Appreciate if you can finalize with @phistuck. Just take into the account the [1st principle here ](https://github.com/i0natan/nodebestpractices/blob/master/writing-guidelines.md) and the fact that we don't have a separate page for code bullets - the home page should stay short and focused

**Comment by sagirk** [comment]
@i0natan Sure.

@phistuck Just [this](https://github.com/i0natan/nodebestpractices/pull/235#issuecomment-418691019) and then we're good to merge.

**Comment by sagirk** [comment]
Hey @phistuck, did you get a chance to give this a shot?

I do understand that sometimes life just gets in the way. If you are finding it difficult to find the time, I can do the change for you. Just let me know.

**Comment by phistuck** [comment]
@sagirk - done.

**Comment by i0natan** [comment]
@sagirk you have permissions to merge, right?

**Comment by sagirk** [comment]
@phistuck Thanks! 🥇 

@i0natan Yup. Gotta resolve the merge conflict first. Thanks for the reminder! ❤️

**Comment by phistuck** [comment]
@i0natan - unless it creates a huge mess for you, I would rather you did not mention me in that "Thank You Notes" section. Thank you.

**Pull request merged**